### PR TITLE
fix #1010

### LINKED ISSUE #1010: Refactor get_accessible_filepath_ids

### PR DESCRIPTION
**Pull request by antgonza**
To solve this issue we can either do a complex SQL command or simply verify per filepath_id and return true/false. For simplicity and to keep a similar-to-previous/detailed-structure decided to go with the latter option.

### CODE REVIEW

File: qiita_db/meta_util.py [hunk 1]
```diff
@@ -48,7 +47,7 @@ def _get_data_fpids(constructor, object_id):
         return {fpid for fpid, _, _ in obj.get_filepaths()}
 
 
-def get_accessible_filepath_ids(user):
+def validate_filepath_access_by_user(user, filepath_id):
     """Gets all filepaths that this user should have access to
```
Comment (josenavas) [inline qiita_db/meta_util.py]: Can you update the one sentence description of the documentation?

File: qiita_db/meta_util.py [hunk 2]
```diff
@@ -48,78 +47,117 @@ def _get_data_fpids(constructor, object_id):
...
+            (SELECT count(*) FROM qiita.analysis_filepath
+             WHERE filepath_id = {0}) AS analysis""".format(filepath_id)
+        TRN.add(sql)
+
+        arid, sid, pid, jid, anid = TRN.execute_fetchflatten()
+
+        # artifacts
+        if arid:
+            # check the public artifacts
+            public_artifacts = qdb.artifact.Artifact.iter_public()
```
Comment (josenavas) [inline qiita_db/meta_util.py]: I think I have not explained myself correctly on the previous comments. As it is currently written here, you're still iterating over all the artifacts. However, at this point you can now which artifacts this have been seen:
```python
sql = """SELECT artifact_id FROM qiita.artifact_filepath WHERE filepath_id = %s"""
```

Similarly you can do for the other entries. You may be able to modify the above call to return the values in an array so you already have them.

File: qiita_db/support_files/populate_test_db.sql [hunk 3]
```diff
@@ -11,7 +11,7 @@ INSERT INTO qiita.qiita_user (email, user_level_id, password, name,
     '$2a$12$gnUi8Qg.0tvW243v889BhOBhWLIHyIJjjgaG6dxuRJkUM8nXG9Efe', 'Dude',
     'Nowhere University', '123 fake st, Apt 0, Faketown, CO 80302',
     '111-222-3344'),
-    ('shared@foo.bar', 3,
+    ('shared@foo.bar', 4,
```
Comment (antgonza) [inline qiita_db/support_files/populate_test_db.sql]: level 3 is superuser and 4 is use, shared user should be regular user so we can actually test that sharing/etc works ... let's see which other tests will fail ...

**Comment by antgonza** [comment]
Ready for another review.

File: qiita_db/meta_util.py [hunk 4]
```diff
@@ -59,78 +58,97 @@ def _get_data_fpids(constructor, object_id):
...
+        elif pid:
+            # the prep access is given by it's artifacts, if the user has
+            # access to any artifact, it should have access to the prep
+            # [0] cause we should only have 1
+            a = qdb.metadata_template.prep_template.PrepTemplate(
+                pid[0]).artifact
+            if (a.visibility == 'public' or a.study.has_access(user)):
+                return True
+            else:
+                for c in a.children:
```
Comment (josenavas) [inline qiita_db/meta_util.py]:
```python
for c in a.descendants.nodes():
```
`children` is only checking at the immediate children but not to the entire descendants.

File: qiita_db/support_files/patches/50.sql [hunk 5]
```diff
@@ -0,0 +1,15 @@
+-- Feb 1, 2017
```
Comment (josenavas) [inline qiita_db/support_files/patches/50.sql]: Please remove this file as I don't think it belongs to this PR.

**Comment by wasade** [comment]
nothing stood out to me, :+1:

**Pull request merged**